python/hsfs/core/tfdata_engine.py

```diff
@@ -141,37 +150,123 @@ def _de_serialize(serialized_example):
             num_parallel_calls=tf.data.experimental.AUTOTUNE,
         )
 
-        if optimize:
+        if process:
+            dataset = dataset.map(
+                lambda value: _process_example(value),
+                num_parallel_calls=tf.data.experimental.AUTOTUNE,
+            )
             dataset = _optimize_dataset(
                 dataset, batch_size, num_epochs, self._is_training
             )
 
         return dataset
 
+    def tf_csv_dataset(
+        self,
+        batch_size=None,
+        num_epochs=None,
+        one_hot_encode_labels=False,
+        num_classes=None,
+        process=False,
+    ):
+        """
+        Reads and returns tf.data.TFRecordDataset object ready to feed tf keras models
+
+        :param batch_size: size of batch, defaults to None
+        :type batch_size: int, optional
+        :param num_epochs: number of epochs to train, defaults to None
+        :type num_epochs: int, optional
+        :param one_hot_encode_labels: if set true then one hot encode lables, defaults to False
+        :type one_hot_encode_labels: boolean, optional
+        :param num_classes: if above true then provide number of target classes, defaults to None
+        :type num_classes: int, optional
+        :param process: if set true  api will optimise tf data read operation, and return feature vector for model
+        with single input, defaults to False
+        :type  process: int, optional
+        :return: tf dataset
+        :rtype: tf.data.TFRecordDataset
+        """
+
+        if self._training_dataset_format != "csv":
+            raise ValueError(
+                "tf_csv_dataset function works only for training datasets that have csv format"
+            )
+
+        if process and (batch_size is None or num_epochs is None):
+            raise ValueError(
+                "if process is set to True you also need to provide batch_size and num_epochs"
+            )
+
+        if one_hot_encode_labels and (num_classes is None or num_classes <= 1):
+            raise ValueError(
+                "if one_hot_encode_labels is set to True you also need to provide num_classes > 1"
+            )
+
+        csv_header = [feat.name for feat in self._training_dataset_schema]
+        record_defaults = [
+            _convert2tfdtype(feat.type) for feat in self._training_dataset_schema
+        ]
+
+        csv_dataset = tf.data.experimental.CsvDataset(
+            self._input_files, header=True, record_defaults=record_defaults,
+        )
+
+        def _process_csv_dataset(csv_record):
+            csv_record_list = list(csv_record)
+            # get target variable 1st
+            y = csv_record_list.pop(csv_header.index(self._target_name))
+            y = tf.convert_to_tensor(y)
+            if one_hot_encode_labels:
+                y = tf.one_hot(y, num_classes)
+            else:
+                y = tf.cast(y, tf.float32)
+
+            # now get the feature vector
```

Comment:
I don't understand why you need to do this. in this example they just pass the types with the header:
```python
titanic_types  = [tf.int32, tf.string, tf.float32, tf.int32, tf.int32, tf.float32, tf.string, tf.string, tf.string, tf.string] 
dataset = tf.data.experimental.CsvDataset(titanic_file, titanic_types , header=True)

for line in dataset.take(10):
  print([item.numpy() for item in line])
```
https://www.tensorflow.org/guide/data#consuming_csv_data

It seems like they do this for you under the hood already.

Reply:
okay, maybe I don't understand the usage of this `process` flag and it's output. we can leave it like that then. I will play around with it, so  I understand better what it's about.